feat: replace cross-chain token standards doc with ERC-7786 focused doc (#151)
defi-wonderland/handbook: sites/ef/docs/interop/erc-7786.md

> # ERC-7786

Reviewer: It would be nice to start with a paragraph explaining with not-so technical definitions what erc-7786 is

Reviewer: Focus on `sendMessage`

Reviewer 2: Makes sense to me

## The timing was not right

As with [ERC-5792 and EIP-7702: a retrospective](./erc-5792-eip-7702.md), the primary developer-facing problem has already been solved. Each bridge already has its own working implementation, and there is little reason to add compatibility with a generalised standard that does not yet have wide adoption. Staying put has no cost, and if this mentality is adopted by the entire ecosystem, the result is a standard with very low adoption.

Every major messaging protocol has already deployed contracts on dozens of chains and onboarded clients. LayerZero supports over [150 chains](https://layerzeroscan.com/). Wormhole, Axelar, Hyperlane and Chainlink CCIP each carved out their own segment with specialized security models and tooling.

From an application developer's perspective, these protocols expose fairly similar interfaces and capabilities. The differences come down to specific features, security models and gas optimizations. Most of their integrations are already in production and some are immutable. Migrating to a new interface would mean disrupting their entire operation for something that offers them no tangible benefit.

> Technically superior standards routinely fail to displace incumbents that work well enough. An example of this can be seen outside of the blockchain space. IPv4 continues to dominate the internet despite IPv6 being available for decades.

Reviewer: I guess IPv6 won where it won because IPv4 became more expensive than migration. What could be the point where ERC-7786 may go through the same situation?

Author: I truly believe that if this is adopted, It will be a slow and forced adoption. The community will become aware of these kinds of standards only after a certain number of negative incidents occur that force them to prioritize these issues.

> ## Different players, different incentives

Reviewer: What do you think about the security model is actually the product. It means that, a standard interface over heterogeneous trust assumptions doesn't unify what app developers actually care about. ERC20 worked because the underlying primitive is just a balance ledger. But cross-chain messaging might not be.

Reviewer 2: You're saying that the standard focuses on the messaging layer, which isn't the main product of interest?

Reviewer: yeah, because 1. no end-users are aware of the messaging layer, but its effects & failure mode likelihoods and 2. there is no major multi-infra cases for the apps to benefit from a common message layer, generally it is from app A to app A', so the benefit is isolated to the same consumer

Beyond inertia, there is an incentive problem.

From an application's perspective, being tied to a single bridge is a risk, and ERC-7786 allows them to have optionality.

> For a bridge, the calculus is different. A cross-chain messaging protocol might prefer to maintain the friction that retains its clients. If migrating to another bridge becomes easy, retaining clients becomes harder.

Reviewer: One interesting thing to consider here is that players adopting ERC-7786 have an incentive to commoditize competitors by adopting this interface, since that their differentiated security model remains *the product*.

Meanwhile, others want to still compete in the "message interface" and tooling, because, if they use ERC-7786 instead, they may lose the race for the security model as a product, being exposed to everyone at the same level (everyone on erc7786).

Author: What I'm wondering is whether that incentive outweighs the potential cost of adopting ERC7786. cc @0xrcinus

Reviewer 2: I think our challenge is that at the moment I don't think (?) we have live examples - maybe a new bridge might adopt it?

Reviewer: A bit unrelated, but if there were a new rollup stack, it should adopt 7786 (hey native rollups)

## Abstraction alone does not drive adoption

> Even if timing and incentives aligned, ERC-7786 faces another problem. It wraps existing bridges but does not add any additional capabilities. Every major messaging protocol already provides its own verification layer.

Reviewer 2: > Every major messaging protocol already provides its own verification layer

what does this mean exactly?

Author: ERC-7786 puts a standard interface on top of existing bridges without adding any new capabilities. Each bridge already has its own security system (verification layer), and ERC-7786 doesn't improve any of them. It just provides portability. 

There was a previous [critique on X](https://x.com/kramnotmark/status/1907508307541295312) about xERC20 and Glashi that raised the same point, and I think it applies here too. cc @Joxess

Previously, aggregation projects like [Glacis](https://www.glacislabs.com/) and [Hashi](https://crosschain-alliance.gitbook.io/hashi) had the same adoption problems. The problem is that a standard interface over already functional solutions does not generate incentives for adoption. On the other hand, it can end up introducing integration risk, as was the case with Glacis and Hashi.

Displacing already functional standards, even if fragmented, must offer real improvements in security and/or incentives that justify the migration.

## What would need to change?

> The lessons learned from ERC-7786 teach us that future standards need the right incentives to achieve adoption. We need to step back and think about how adoption actually gets generated before working on any standard. ERC-7786 attempted to carve out a place in a space where incumbents already offered functional solutions tailored to each consumer’s preferred security model at low gas cost. Bringing only abstraction in return won't be enough to generate adoption. That standard is going to have a bad time.

Reviewer 2: ```suggestion
ERC-7786 attempted to carve out a place in a space where incumbents already provide sufficiently-functional solutions to entrenched users. Offering only abstraction wasn't enough to convince bridges and asset issuers - another reminder that standards can't take adoption for granted.
```

Author: I think we can find a way to incorporate what Joxes says in the comment about this paragraph and mention that the incentives need to be right, or there won't be any adoption. (Maybe forced adoption in the long run)

> The lessons learned from ERC-7786 teach us that future standards need the right incentives to achieve adoption. We need to step back and think about how adoption actually gets generated before working on any standard. ERC-7786 attempted to carve out a place in a space where incumbents already offered functional solutions tailored to each consumer’s preferred security model at low gas cost. Bringing only abstraction in return won't be enough to generate adoption. That standard is going to have a bad time.

Reviewer: I'm about to say this (it sounds awful, but it's the reality)... Only when things go wrong (for example, a bridge unfortunately suffering an exploit or being forced to wind down) does ERC7786 reinforce its reason for existing, since it's the type of event that most often causes migrations of different kinds. It is probably the main driver today.

When things are stable, no one thinks about using a standard for the purpose of changing the bridge

Author: @0xrcinus It would make sense to add this at the end of the article. I also think that it is probably the main driver today. wdyt?

Reviewer 2: I generally agree that adoption of these kind of standards will come from shocks or new entrants

# ERC-7786

## Useful in theory, unused in practice

[ERC-7786](https://eips.ethereum.org/EIPS/eip-7786) defines a standard interface for cross-chain messaging. Its core primitive is a single function called `sendMessage` that takes a destination chain, a recipient and a message. The bridge handles delivery. If every bridge exposed this same API, developers could switch providers without rewriting their contracts.

This sounds like a useful primitive, but the standard has not gained traction. The following sections explore why.

## The timing was not right

As with [ERC-5792 and EIP-7702: a retrospective](./erc-5792-eip-7702.md), the primary developer-facing problem has already been solved. Each bridge already has its own working implementation, and there is little reason to add compatibility with a generalised standard that does not yet have wide adoption. Staying put has no cost, and if this mentality is adopted by the entire ecosystem, the result is a standard with very low adoption.

> Every major messaging protocol has already deployed contracts on dozens of chains and onboarded clients. LayerZero supports over [150 chains](https://layerzeroscan.com/). Wormhole, Axelar, Hyperlane and Chainlink CCIP each carved out their own segment with specialized security models and tooling.

Reviewer: Should we mention that even rollup architectures do the same

Author: I'm not sure about this. I think we could leave it as is, without mentioning the rollups.
I'm open to whatever you both prefer.

> From an application developer's perspective, these protocols expose fairly similar interfaces and capabilities. The differences come down to specific features, security models and gas optimizations. Most of their integrations are already in production and some are immutable. Migrating to a new interface would mean disrupting their entire operation for something that offers them no tangible benefit.

Reviewer: "Some are immutable", also the apps that are serving

Author: Yes 100%, it was in the early versions in another section, but we ended up removing it since it would look even cleaner this way.

## Abstraction alone does not drive adoption

Even if timing and incentives aligned, ERC-7786 faces another problem. It wraps existing bridges but does not add any additional capabilities. Every major messaging protocol already provides its own verification layer.

Previously, aggregation projects like [Glacis](https://www.glacislabs.com/) and [Hashi](https://crosschain-alliance.gitbook.io/hashi) had the same adoption problems. The problem is that a standard interface over already functional solutions does not generate incentives for adoption. On the other hand, it can end up introducing integration risk, as was the case with Glacis and Hashi.

Displacing already functional standards, even if fragmented, must offer real improvements in security and/or incentives that justify the migration.

## What would need to change?

ERC-7786 attempted to carve out a place in a space where incumbents already provide sufficiently-functional solutions to entrenched users. Offering only abstraction wasn't enough to convince bridges and asset issuers - another reminder that standards can't take adoption for granted.

> The status quo for message passing is unlikely to change, without some external shock. This could be triggered by events, such as the recent [KelpDAO and LayerZero incident](https://layerzero.network/blog/kelpdao-incident-statement), which might prompt incumbents to revisit their security assumptions. Or it could be driven by new entrants shaking up the market, as more assets and activity come onchain. In either case, the adoption of ERC-7786 will be determined by markets more than working groups.

Reviewer: https://github.com/defi-wonderland/handbook/pull/151/changes#r3288830248 so I think Ethereum has one last shot to pump the standard

Previously, aggregation projects like [Glacis](https://www.glacislabs.com/) and [Hashi](https://crosschain-alliance.gitbook.io/hashi) had the same adoption problems. The problem is that a standard interface over already functional solutions does not generate incentives for adoption. On the other hand, it can end up introducing integration risk, as was the case with Glacis and Hashi.

Displacing already functional standards, even if fragmented, must offer real improvements in security and/or incentives that justify the migration.

## What would need to change?

ERC-7786 attempted to carve out a place in a space where incumbents already provide sufficiently-functional solutions to entrenched users. Offering only abstraction wasn't enough to convince bridges and asset issuers - another reminder that standards can't take adoption for granted.

But this isn't unique to ERC-7786. The ecosystem seems to move reactively, learning mainly from frictions, failures and real incidents. It's only when negative events come with massive losses, like [the KelpDAO and LayerZero exploit](https://layerzero.network/blog/kelpdao-incident-statement), that the ecosystem starts taking security seriously.

> Not all hope is lost. This standard may still find its moment. Initiatives like [ERC-7965](https://eips.ethereum.org/EIPS/eip-7965), which adds a layer of cryptographic verification on top of ERC-7786, could make the standard attractive to new entrants. Applications building crosschain products could have an incentive to push the bridges they rely on to integrate it as an additional security measure.

Reviewer 2: Do you think 7965 will be enough to drive adoption of 7786? That seems like a stretch to me. And this paragraph ends a bit abruptly, I'm not sure it's the best final note. I think there are two hopeful angles:
- Security incidents causing teams to reassess their security assumptions
- New entrants who might introduce new approaches, leveraging open standards
I think it is reasonable to end with those broader points, rather than this ERC, what do you think @Joxess

Author: Tbh, I don't think 7965 alone will be enough. I tried to sound positive. Agreed on both angles. They are more clear. 
Sending a modified version now

> Not all hope is lost. This standard may still find its moment. Initiatives like [ERC-7965](https://eips.ethereum.org/EIPS/eip-7965), which adds a layer of cryptographic verification on top of ERC-7786, could make the standard attractive to new entrants. Applications building crosschain products could have an incentive to push the bridges they rely on to integrate it as an additional security measure.

Author: Thoughts on this version @0xrcinus @Joxess 

```suggestion
What's true is that standards rarely succeed on their own merits alone. ERC-7786 could find its place through security incidents such as the [KelpDAO and LayerZero exploit](https://layerzero.network/blog/kelpdao-incident-statement) that push incumbents to revisit their security assumptions. On the other hand, new entrants introducing innovative approaches could adopt it to lower the integration barrier for apps.
```

Reviewer 2: ```suggestion
The status quo for message passing is unlikely to change, without some external shock. This could be triggered by events, such as the recent [KelpDAO and LayerZero incident](https://layerzero.network/blog/kelpdao-incident-statement), which might prompt incumbents to revisit their security assumptions. Or it could be driven by new entrants shaking up the market, as more assets and activity come onchain. In either case, the adoption of ERC-7786 will be determined by markets more than working groups.
```
